src/plugins/internalnotification/macros/addType.h

```diff
@@ -0,0 +1,71 @@
+#ifndef TYPE
+#error "You have to #define TYPE, TYPE_NAME and TO_VALUE before including the addType supermacro"
+#endif
+#ifndef VALUE_TYPE
+// use type as default if not set
+#define VALUE_TYPE TYPE
+#endif
+#ifndef TYPE_NAME
+#error "You have to #define TYPE, TYPE_NAME and TO_VALUE before including the addType supermacro"
+#endif
+#ifndef TO_VALUE
+#error "You have to #define TYPE, TYPE_NAME and TO_VALUE before including the addType supermacro"
+#endif
+#ifndef CHECK_CONVERSION
+#define CHECK_CONVERSION 1
+#endif
+
+#define CONCAT(X, Y) CONCAT2 (X, Y)
+#define CONCAT2(X, Y) X##Y
+
+#define INTERNALNOTIFICATION_CONVERSION_CALLBACK_NAME(TYPE_NAME) CONCAT (elektraInternalnotificationConvert, TYPE_NAME)
+
+#define INTERNALNOTIFICATION_REGISTER_SIGNATURE(TYPE, TYPE_NAME)                                                                           \
+	int INTERNALNOTIFICATION_REGISTER_NAME (TYPE_NAME) (Plugin * handle, Key * key, TYPE * variable)
+
+#define INTERNALNOTIFICATION_CONVERSION_CALLBACK_SIGNATURE(TYPE_NAME)                                                                      \
+	void INTERNALNOTIFICATION_CONVERSION_CALLBACK_NAME (TYPE_NAME) (Key * key, void * context)
+
+INTERNALNOTIFICATION_CONVERSION_CALLBACK_SIGNATURE (TYPE_NAME)
+{
+	TYPE * variable = (TYPE *) context;
+	char * end ELEKTRA_UNUSED;
+	const char * string = keyValue (key);
+	errno = 0;
+	/* convert string to target type */
+	VALUE_TYPE value = TO_VALUE;
+	/* only update if conversion was successful */
+	if (CHECK_CONVERSION)
+	{
+		*(variable) = value;
+	}
+	else
+	{
+		ELEKTRA_LOG_WARNING ("conversion failed! string=%s, stopped=%c errno=%d", keyString (key), *end, errno);
```

Comment:
So the value is silently (unless the logger is enabled) not changed?

Not so ideal...

Reply:
> (e.g. the type plugin) to allow only valid values the key database.

Yes, that is the goal. But there is no type validation for C types.

> elektraNotificationSetConversionFailedCallback

It could be useful for applications to log and abort. But it is much better if such situations cannnot occur :-)